File: LeetCode/BinaryTreeDFS/JavaScript/symmetricTree.js
// Given the root of a binary tree, check whether it is a mirror of itself (i.e., symmetric around its center).

// Example 1:

// Input: root = [1,2,2,3,4,4,3]
// Output: true
// Example 2:

// Input: root = [1,2,2,null,3,null,3]
// Output: false

// Constraints:

// The number of nodes in the tree is in the range [1, 1000].
// -100 <= Node.val <= 100

// Follow up: Could you solve it both recursively and iteratively?

// TC: O(n) where n is the number of nodes. We'll visit all of them
// SC: O(h) where h is the height of the tree.

const Result = { PASS: "\x1b[92mPASS\x1b[0m", FAIL: "\x1b[91mFAIL\x1b[0m" };

/**
 * Definition for a binary tree node.
 */
class TreeNode {
  constructor(val, left, right) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

class SymmetricTreeRecord {
  constructor(values, expected) {
    this.values = values;
    this.expected = expected;
  }
}

class Solution {
  /**
   * @param {TreeNode} root
   * @return {boolean}
   */
  isSymmetric1(root) {
    // Using Recursive DFS
    // Declare a dfs function
    function DFS(lNode, rNode) {
      if (!lNode && !rNode) return true; // both are null, they are symetric
      if (!lNode || !rNode) return false; // either node is null... we've already check prior that both are not null. Not symmetric
      // Check the values of the nodes are equal and perform DSF inner nodes and outer nodes
      return (
        lNode.val === rNode.val &&
        DFS(lNode.left, rNode.right) &&
        DFS(lNode.right, rNode.left)
      );
    }

    return DFS(root.left, root.right);
  }

  /**
   * @param {TreeNode} root
   * @return {boolean}
   */
  isSymmetric2(root) {
    if (!root) {
      return true;
    }

    // Using Iterative DFS
    let stack = [];

    stack.push([root.left, root.right]);

    while (stack.length) {
      let [nodeL, nodeR] = stack.pop();

      if (!nodeL && !nodeR) {
        continue;
      } else if (!nodeL || !nodeR) {
        return false;
      } else if (nodeL.val !== nodeR.val) {
        return false;
      } else {
        // Add the outer and inner nodes to the stack
        stack.push([nodeL.left, nodeR.right]);
        stack.push([nodeL.right, nodeR.left]);
      }
    }

    return true;
  }
}

function createTree(values) {
  if (!values.length || values[0] === null) return null;

  const root = new TreeNode(values[0]);
  const queue = [root];
  let index = 1;

  while (queue.length && index < values.length) {
    const node = queue.shift();

    if (values[index] !== null && values[index] !== undefined) {
      node.left = new TreeNode(values[index]);
      queue.push(node.left);
    }
    index++;

    if (index < values.length && values[index] !== null && values[index] !== undefined) {
      node.right = new TreeNode(values[index]);
      queue.push(node.right);
    }
    index++;
  }

  return root;
}

function testSolution(record) {
  const solution = new Solution();
  const result = solution.isSymmetric2(createTree(record.values));
  const pass = result === record.expected;

  console.log(`Input: root = ${JSON.stringify(record.values)}`);
  console.log(`Expected: ${record.expected}`);
  console.log(`Result: ${result}`);
  console.log(pass ? Result.PASS : Result.FAIL);
}

const records = [
  new SymmetricTreeRecord([1, 2, 2, 3, 4, 4, 3], true),
  new SymmetricTreeRecord([1, 2, 2, null, 3, null, 3], false),
  // A single node tree is always a mirror of itself
  new SymmetricTreeRecord([1], true),
  // The children mirror each other even though their own children are missing on opposite sides
  new SymmetricTreeRecord([1, 2, 2, null, 3, 3, null], true),
  // The children values differ, so the tree is not symmetric
  new SymmetricTreeRecord([1, 2, 3], false),
];

records.forEach((record, index) => {
  console.log(`# Test case ${index + 1}`);
  testSolution(record);
  console.log("----------------------------------------");
});

File: LeetCode/BinaryTreeDFS/python/symmetricTree.py
# Definition for a binary tree node.
from collections import deque
from typing import Dict, List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        # Use BFS to check if left child is the same as right child
        if root is None:
            return True
        
        queue = deque([]) # initialize deque with empty list
        queue.append([root.left, root.right]) # add the pair of children as a list => [[child.left, child.right]]

        while len(queue) > 0 :
            qlen = len(queue)

            for _ in range(qlen):
                [nodeL, nodeR] = queue.popleft()

                if nodeL is None and nodeR is None:
                    continue
                elif nodeL is None or nodeR is None:
                    return False
                elif nodeL.val != nodeR.val:
                    return False
                else:
                    queue.append([nodeL.left, nodeR.right])
                    queue.append([nodeL.right, nodeR.left])
        
        return True

    @staticmethod
    def testSolution(input: Dict[str, List[str|int]]) -> None:
        # Create the Binary Tree from the input list using BFS
        binTree = Solution.BFSCreateBinaryTree(input["root"])

        # Print the Binary Tree and the expected output
        print("Input: " + Solution.BFSBinaryTreeToStr(binTree))
        print("Expected: " + str(input["expected"]))

        # Call the function to be tested
        res = Solution().isSymmetric(binTree)

        # Print the result
        print("Result: " + str(res))

        # Validate the result
        print("PASS" if res == input["expected"] else "FAIL")
        print("-" * 50)

    @staticmethod
    def BFSCreateBinaryTree(node: List[int|str]) -> TreeNode:
        #  0 1 2  3    4    5  6
        # [3,9,20,null,null,15,7]
        #           3
        #        9     20
        #            15   7

        arLen = len(node)

        if arLen < 1:
            return None

        root = TreeNode(node[0], None, None)
        i = 1

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])

        while len(queue) > 0 and i < arLen:
            # Pre order: root, left, then right
            cur_node = queue.popleft()

            # Left child
            if i < arLen and node[i] != "null":
                cur_node.left = TreeNode(node[i], None, None)
                queue.append(cur_node.left)
            i += 1
            
            # Right child
            if i < arLen and node[i] != "null":
                cur_node.right = TreeNode(node[i], None, None)
                queue.append(cur_node.right)            
            i += 1

        return root
    
    @staticmethod
    def BFSBinaryTreeToStr(root: Optional[TreeNode]) -> None:
        if root is None:
            return "[ ]"
        
        sb = []

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])


        while len(queue) > 0:
            curNode = queue.popleft()

            tmp = str(curNode.val) if curNode is not None else "null"

            sb.append(tmp)
 
            if curNode is not None:
                queue.append(curNode.left)
                queue.append(curNode.right)

        new_len = 0

        # Check starting from end of sb_array where the value is not "null"
        for i in range(len(sb) - 1, -1, -1):
            if sb[i] == "null":
                continue
            else:
                new_len = i
                break

        # Remove those extra "null"
        if len(sb) > 0:
            sb = sb[0 : new_len + 1]

        return "[ " + ", ".join(sb) + " ]"

if __name__ == "__main__":
    input = {"root": [1,2,2,3,4,4,3], "expected": True}
    Solution.testSolution(input)

    input = {"root": [1,2,2,"null",3,"null",3], "expected": False}
    Solution.testSolution(input)